src/storage-controller/src/instance.rs

```diff
@@ -132,18 +148,66 @@ where
         let metrics = self.metrics.for_replica(id);
         let replica = Replica::new(id, config, self.epoch, metrics, self.response_tx.clone());
 
+        self.replicas.insert(id, replica);
+
+        self.update_ingestion_scheduling(false);
+
+        self.replay_commands(id);
+    }
+
+    /// Replays commands to the specified replica.
+    pub fn replay_commands(&mut self, replica_id: ReplicaId) {
+        let commands = self.history.iter().cloned().collect::<Vec<_>>();
```

Comment:
Haven't tried it but it feels like the `collect` and the `into_iter` below can be deleted.